base/multidimensional.jl

```diff
@@ -1064,6 +1064,213 @@ end
     $(_generate_unsafe_setindex!_body(2))
 end
 
+# Reduction support for CartesianIndices
+halves(inds::CartesianIndices) = map(CartesianIndices∘tuple, _halves(inds.indices)...)
+function mapreduce_kernel(f, op, A, init, inds::CartesianIndices{N}) where {N}
+    N == 0 && return _mapreduce_start(f, op, A, init, A[inds[]])
+    N == 1 && return mapreduce_kernel(f, op, A, init, inds.indices[1])
+    is_commutative_op(op) && return mapreduce_kernel_commutative(f, op, A, init, inds)
+    r = inds.indices[1]
+    if length(r) == 1
+        # A one-element inner loop is less-than-helpful
+        i1, s = iterate(inds)
+        a1 = @inbounds A[i1]
+        v = _mapreduce_start(f, op, A, init, a1)
+        for i in Iterators.rest(inds, s)
+            ai = @inbounds A[i]
+            v = op(v, f(ai))
+        end
+    else
+        # Divide the loop into an outer and inner loop, and skip the first element
+        # in the first iteration of the outer loop
+        outer = CartesianIndices(tail(inds.indices))
+        o1, so = iterate(outer)
+        v = _mapreduce_start(f, op, A, init, A[r[begin], o1])
+        for i in r[begin+1:end]
+            ai = @inbounds A[i, o1]
+            v = op(v, f(ai))
+        end
+        for o in Iterators.rest(outer, so)
+            for i in r
+                ai = @inbounds A[i, o]
+                v = op(v, f(ai))
+            end
+        end
+    end
+    return v
+end
+
+function mapreduce_kernel_commutative(f, op, A, init, inds::AbstractArray)
+    if length(inds) < 16
+        i1, iN = firstindex(inds), lastindex(inds)
+        v_1 = _mapreduce_start(f, op, A, init, @inbounds A[inds[i1]])
+        for i in i1+1:iN
+            a = @inbounds A[inds[i]]
+            v_1 = op(v_1, f(a))
+        end
+        return v_1
+    end
+    return _mapreduce_kernel_commutative(f, op, A, init, inds)
+end
+
+# This special internal method must have at least 8 indices and allows passing
+# optional scalar leading and trailing dimensions
+function _mapreduce_kernel_commutative(f, op, A, init, inds, leading=(), trailing=())
```

Comment:
potentially profitable to unroll it out to `@nexprs 16` when on a 64bit platform but `sizeof(T) <= 4`